src/linker/Linker/CompilerGeneratedNames.cs

```diff
@@ -0,0 +1,70 @@
+using System.Diagnostics.CodeAnalysis;
+
+namespace Mono.Linker
+{
+	class GeneratedNames
+	{
+		internal static bool IsGeneratedMemberName (string memberName)
+		{
+			return memberName.Length > 0 && memberName[0] == '<';
+		}
+
+		internal static bool IsLambdaDisplayClass (string className)
+		{
+			if (!IsGeneratedMemberName (className))
+				return false;
+
+			// This is true for static lambdas (which are emitted into a class like <>c)
+			// and for instance lambdas (which are emitted into a class like <>c__DisplayClass1_0)
+			return className.StartsWith ("<>c");
+		}
+
+		// Lambda methods have generated names like "<UserMethod>c__0_1" where "UserMethod" is the name
```

Comment:
Did we consider if this approach is robust enough?
I know we have to rely on detection of compiler generated names, but I honestly didn't expect to parse the name to get information from it this way. The current approach can't differentiate overloads, so what might happen is this:
```C#
[UnconditionalSuppressMessage("IL2026")]
void TestMethod()
{
}

void TestMethod(string p)
{
    CallLocal ();

    void CallLocal()
    {
        MethodWithRUC();  // No warning - since we would apply the suppression from TestMethod() onto local function from TestMethod(string)
    }
}

[RequiresUnreferencedCode("Don't call me")]
void MethodWithRUC() { }
```

I thought we would have to scan IL of all methods on the containing type to find the method which refers to the lambda/local in question and associate them that way. It is potentially relatively expensive, but then again it should only happen on types which use lambdas/local functions which should be a relative minority.

Reply:
This was just a quick and dirty way to hook it up to test that this would correctly identify the lambdas and local functions. Thanks for the example - you are right that it will probably need to actually look at IL references rather than the name - I'll make sure that case is tested.

Reply:
I was just browsing the repo... only noticed this is WIP after I posted the comment.